### PR TITLE
[Merged by Bors] - refactor(AlgebraicGeometry/EllipticCurve/Weierstrass): change the variableChange into a structure

### PR DESCRIPTION
**Pull request by acmepjz**
Main changes:

- [x] Change the `variableChange` (was `u r s t` everywhere) into a structure `VariableChange`.
- [x] Add `id`, `comp` and `inv` to `VariableChange` and prove that it form a group and acts on the set of elliptic curves to the left.
- [x] Add `baseChange` to `VariableChange`, prove the compatibility of it with base change of elliptic curves.
- [x] Prove that if the ring homomorphism is injective, then `baseChange` of `VariableChange` and elliptic curves are injective (NB: We don't say anything about elliptic curves modulo isomorphism yet!)
- [x] Prove that `baseChange` of `VariableChange` is a group homomorphism

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:38]: "change of variable*s*"? I've been using the plural form everywhere below. I'm also unsure if we should call it `VariableChange` since it might be confused with `variableChange`, but I don't have a better name either.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:189]: "change of variable*s*" plural form

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:212]: My convention is to have as many terms as possible in the same line, so e.g. `- (C.t + C.r * C.s) * W.a₁` (and maybe even `+ 3 * C.r ^ 2`) should be in the same line, and then line break afterwards

**Review by Multramate — APPROVED** [review]
Looks good to me

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:38]: Yes, it's called change of variable**s** in Silverman's book. I think the function to change the curve could be called `changeCurve` (at least in my previous attempt).

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: What about `changeCoordinate`? I'm trying to compare it with `baseChange` (which should probably be called `changeBase` if we decide to rename everything), where it's clear what we're changing (whereas `changeCurve` doesn't say exactly what).

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: My proposal `changeCurve` is in fact coming from the names of corresponding functions in Pari/GP: `ellchangecurve()`, `ellchangepoint()` and `ellchangepointinv()`.
```
ellchangecurve(E,v):

   Changes  the data for the elliptic curve E by changing the coordinates using
the vector v = [u,r,s,t],  i.e. if x' and y' are the new coordinates,  then x =
u^2x'+r, y = u^3y'+su^2x'+t.   E must be an ell structure as output by ellinit.
The  special case v = 1 is also used instead of [1,0,0,0] to denote the trivial
coordinate change.

ellchangepoint(x,v):

   Changes  the coordinates of the point or vector of points x using the vector
v = [u,r,s,t], i.e. if x' and y' are the new coordinates, then x = u^2x'+r, y =
u^3y'+su^2x'+t (see also ellchangecurve).

   ? E0 = ellinit([1,1]); P0 = [0,1]; v = [1,2,3,4];
   ? E = ellchangecurve(E0, v);
   ? P = ellchangepoint(P0,v)
   %3 = [-2, 3]
   ? ellisoncurve(E, P)
   %4 = 1
   ? ellchangepointinv(P,v)
   %5 = [0, 1]
```
BTW, I think `changeRing` is better than `changeBase` (which sounds too weird), while I think `baseChange` is the best, since it is a standard word in mathematics.

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: I see - let's do `changeCurve` and `changeRing` then

**Comment by acmepjz** [comment]
Done with `changeCurve`. For `changeRing` I think it's better in another PR?

PS: I noticed that `changeCurve` and `changePoint` are member functions of `WeierstrassCurve` and `Point`, respectively, unlike my design that they are member functions of `VariableChange`. So technically they can all share the same name `variableChange`.

**Comment by Multramate** [comment]
> Done with `changeCurve`. For `changeRing` I think it's better in another PR?
> 
> PS: I noticed that `changeCurve` and `changePoint` are member functions of `WeierstrassCurve` and `Point`, respectively, unlike my design that they are member functions of `VariableChange`. So technically they can all share the same name `variableChange`.

Yeah, that was kinda my original hesitance. I definitely don't want to have it as member functions of something else other than `WeierstrassCurve`, so that we can use dot notation `E.variableChange`, and it sounded weird saying `E.changeCurve` (we're taking a curve and then changing the curve?). You've convinced me with the existing function in Pari/GP, but now that you've mentioned the `Point` version it's gonna be even weirder with `P.changeCurve` (we're taking a point and then changing the curve?), whereas `changeCoordinate` or `variableChange` is understood for both the curve and the point. I leave it up to you to decide whether we want to revert to `variableChange` or leave the changes as it is, but in any case I think renaming `baseChange` can be done in the same PR (and we'll put it in the commit description). Sorry for the trouble!

**Comment by acmepjz** [comment]
> You've convinced me with the existing function in Pari/GP, but now that you've mentioned the `Point` version it's gonna be even weirder with `P.changeCurve` (we're taking a point and then changing the curve?), whereas `changeCoordinate` or `variableChange` is understood for both the curve and the point.

For the point it should be `P.changePoint` 😂.

> Yeah, that was kinda my original hesitance. I definitely don't want to have it as member functions of something else other than `WeierstrassCurve`, so that we can use dot notation `E.variableChange`, and it sounded weird saying `E.changeCurve` (we're taking a curve and then changing the curve?).

Ah, I remembered the design consideration in my previous implementation. It uses `C.changeCurve E` and `C.changePoint P`. Whereas in current mathlib4 version, it uses `E.changeCurve C` and `P.changePoint C` (to be renamed with `variableChange`).

My design emphasis the fact that change of variables form a group and acts on the set of elliptic curves to the left. (It also acts on the set of points, but I forgot if it's also acts to the left.) Namely, there is the identity change, the composite of change `C1.composite C2`, etc etc., which satisfies `(C1.composite C2).changeCurve E = C1.changeCurve (C2.changeCurve E)`, etc etc.
I think the advantage is that it allows us to define the automorphism group of an elliptic curve easily.

I think this will introduce significant changes to the current implementation. What's your opinion?

**Comment by Multramate** [comment]
Mathlib certainly has right actions e.g. [rightMul](https://leanprover-community.github.io/mathlib4_docs//Mathlib/Algebra/Group/Defs.html#rightMul) and I don't know what the usual convention in the literature is (my hunch is it's acting on the left). These are good to have in mind, but ultimately we want to establish a sensible and consistent API that's easy to use for other applications, and my opinion is that we should stick to the `WeierstrassCurve` namespace (as the main subject we're dealing with is the Weierstrass curve). For instance, I think it still makes sense to say `E.changeCurve (C1.composite C2)`, and we can still define the automorphism set of `E` as `{C : VariableChange // E.changeCurve C = E}` rather than making it a structure. If the action direction gets confusing we can always clarify with docstrings.

**Comment by acmepjz** [comment]
OK, let's stick to current design then. I choose to come back to `variableChange` and leave `baseChange` unchanged, since renaming it require too many changes in both of these two files.

**Comment by acmepjz** [comment]
Should I add more properties of `VariableChange` to this PR, e.g. they form a group? Or should I open a new PR later?

**Comment by Multramate** [comment]
I think doing it here is fine, since the maintainers are currently busy with the port and probably won't look at this just yet.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:254]: Golf with `linear_combination` and `Units.inv_mul`:
```suggestion
lemma comp_left_inv (C : VariableChange R) : comp (inv C) C = id := by
  rw [comp, id, inv]
  ext
  · exact C.u.inv_mul
  · linear_combination (norm := (simp only; ring1)) -C.r * pow_mul_pow_eq_one 2 C.u.inv_mul
  · linear_combination (norm := (simp only; ring1)) -C.s * C.u.inv_mul
  · linear_combination (norm := (simp only; ring1))
      (C.r * C.s - C.t) * pow_mul_pow_eq_one 3 C.u.inv_mul
        + -C.r * C.s * pow_mul_pow_eq_one 2 C.u.inv_mul
```
You can change back to `one` and `mul` if you prefer.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:264]: Golf with `<;>`:
```suggestion
lemma comp_assoc (C C' C'' : VariableChange R) : comp (comp C C') C'' = comp C (comp C' C'') := by
  ext <;> simp only [comp, Units.val_mul] <;> ring1
```
You can change back to `one` and `mul` if you prefer.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:207]: I guess `⟨1, 0, 0, 0⟩` is probably fine since they're so short?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:218]: You want docstrings on `def`s. They're not checked in the CI linter currently for the porting but you can check them locally with `#lint`, and I think they'll add this back eventually.

**Review by Multramate — COMMENTED** [review]
I've left some comments, mostly stylistic but with some golfs

**Review by Multramate — APPROVED** [review]

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: > I need to write `smul := fun C W => variableChange W C` and then `one_smul := variableChange_id` refuses to work

Oops, it works. Nevermind.

**Comment by acmepjz** [comment]
I think the following are also worth to be added:

* base change of `VariableChange`
  * if `B` is an `A`-algebra, then `VariableChange A -> VariableChange B` is a group homomorphism, and if `A -> B` is injective, then this is also injective
  * base change of `VariableChange` compatible with base change of elliptic curves
* `Isomorphism E E'` and they form a groupoid (do we actually need such advanced things? Or only `Isomorphism E E` form a group)
  * if `B` is an `A`-algebra, then there is `Isomorphism E E' -> Isomorphism E/B E'/B`, and if `A -> B` is injective, then this is also injective

The intention is, in another PR, prove that over a field, `Isomorphism E E'` is finite (by proving a complete classification of `Isomorphism E E` over an algebraic closed field).

Any comments?

**Comment by Multramate** [comment]
Yes, I think they're all wonderful! I don't know anything about the groupoid API though, and will be more than happy with just `Isomorphism E E` being a group. The `VariableCHange` stuff can be added in this PR probably, while the rest in another PR.

**Comment by acmepjz** [comment]
There is groupoid API `Mathlib.CategoryTheory.Groupoid` but it involves category, which is overkill I think.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:454]: Surely this is OK:
```suggestion
def baseChange : VariableChange A :=
  ⟨Units.map (algebraMap R A) C.u, algebraMap R A C.r, algebraMap R A C.s, algebraMap R A C.t⟩
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:85]: If you added a bunch of lemmas to `map_simp` for `comp`, you might as well do the same for `id`. On the other hand I personally wouldn't add those lemmas to `map_simp` though, since they're only specific to `Units` or `MonoidHom`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:472]:
```suggestion
lemma baseChange_self : C.baseChange R = C :=
  rfl
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:445]:
```suggestion
lemma baseChange_injective (h : Function.Injective <| algebraMap R A) :
    Function.Injective <| baseChange (R := R) (A := A) := fun W W' h1 => by
  rcases mk.inj h1 with ⟨_, _, _, _, _⟩
  ext <;> apply_fun _ using h <;> assumption
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:486]:
```suggestion
lemma baseChange_injective (h : Function.Injective <| algebraMap R A) :
    Function.Injective <| baseChange (R := R) A := fun C C' h1 => by
  rcases mk.inj h1 with ⟨h1, _, _, _⟩
  replace h1 := (Units.mk.inj h1).left
  ext <;> apply_fun _ using h <;> assumption
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:1192]:
```suggestion
lemma baseChange_injective (h : Function.Injective <| algebraMap R A) :
    Function.Injective <| baseChange (R := R) (A := A) := fun E E' h1 => by
  rcases mk.inj h1 with ⟨h1, h2⟩
  replace h2 := (Units.mk.inj h2).left
  rcases WeierstrassCurve.mk.inj h1 with ⟨_, _, _, _, _⟩
  ext <;> apply_fun _ using h <;> assumption
```

**Review by Multramate — COMMENTED** [review]
Some golfs! I still don't quite get what you mean by blatantly wrong: if we're working over a general field we can still do base and variable changes. Even though the resulting curve may not be isomorphic to the original curve, wouldn't these still be well-defined?

**Comment by acmepjz** [comment]
> I still don't quite get what you mean by blatantly wrong: if we're working over a general field we can still do base and variable changes. Even though the resulting curve may not be isomorphic to the original curve, wouldn't these still be well-defined?

Ah yes, they are well-defined. I just mean that the lemmas only responsible for `EllipticCurve R -> EllipticCurve A` being injective, but not for `EllipticCurve R / isomorphisms -> EllipticCurve A / isomorphisms`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:85]: Oh OK. Let me revert it to the previous state.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: Here I'd like to call `WeierstrassCurve.baseChange_injective` but I can't figure out how to tell `ext` don't expand everything.

**Review by riccardobrasca — COMMENTED** [review]
Can you please add a short description of this PR? Thanks!

**Comment by riccardobrasca** [comment]
LGTM, thanks!

bors merge

**Pull request closed without merging**
